// mini-framework/src/EventManager.js
//more can be added here easily, but need to update DomAbstraction.js as well
export const events = {
    CLICK: "click",
    KEYDOWN: "keydown",
    KEYUP: "keyup",
    MOUSEENTER: "mouseenter",
    SCROLL: "scroll",
    CHANGE: "change",
}

/**
 * Generic function to add event listener to an HTML node element
 * @param {string} eventType - The type of event to handle (e.g., 'click' or 'change').
 * @param {HTMLElement} element - The HTML element to attach the event to
 * @param {function} callback - The event handler function to execute when the event occurs.
 */

export function onEvent(eventType, element, callback) {
    if (callback instanceof Function) {
        element.addEventListener(eventType, callback);
    } else {
        console.error(`Invalid callBack provided, ${callback} is not a function`);
    }
}

/**
 * Triggers a custom event on the specified element
 *
 * @param {string} eventType - The type of custom event to trigger
 * @param {HTMLElement} element - The HTML element on which the custom event is triggered
 * @param {Object} eventDetail - Optional detail to pass with the custom event
 */
export function triggerCustomEvent(eventType, element, eventDetail = null) {
    const customEvent = new CustomEvent(eventType, { detail: eventDetail });
    element.dispatchEvent(customEvent);
}
